ImperatorToCK3/Source/CK3/CK3World.cpp

```diff
@@ -0,0 +1,170 @@
+				else
+					provinces.insert(std::pair(id, provinces.find(newProvince.second)->second));
+			}
+		}
+		catch (std::exception& e)
+		{
+			Log(LogLevel::Warning) << "Invalid province filename: " << ck3Path << "/game/history/province_mapping/" << fileName << " : " << e.what();
+		}
+	}
+	LOG(LogLevel::Info) << ">> Loaded " << provinces.size() << " province definitions.";
+}
+
+void CK3::World::importImperatorProvinces(const ImperatorWorld::World& sourceWorld)
+{
+	LOG(LogLevel::Info) << "-> Importing Imperator Provinces";
+	auto counter = 0;
+	// Imperator provinces map to a subset of CK3 provinces. We'll only rewrite those we are responsible for.
+	for (const auto& province : provinces)
+	{
+		const auto& impProvinces = provinceMapper.getImperatorProvinceNumbers(province.first);
+		// Provinces we're not affecting will not be in this list.
+		if (impProvinces.empty())
+			continue;
+		// Next, we find what province to use as its initializing source.
+		const auto& sourceProvince = determineProvinceSource(impProvinces, sourceWorld);
+		if (!sourceProvince)
+		{
+			continue; // MISMAP, or simply have mod provinces loaded we're not using.
+		}
+		else
+		{
+			province.second->initializeFromImperator(sourceProvince->second, cultureMapper, religionMapper);
+		}
+		// And finally, initialize it.
+		counter++;
+	}
+	LOG(LogLevel::Info) << ">> " << sourceWorld.getProvinces().size() << " Imperator provinces imported into " << counter << " CK3 provinces.";
+}
+
+std::optional<std::pair<int, std::shared_ptr<ImperatorWorld::Province>>> CK3::World::determineProvinceSource(const std::vector<int>& impProvinceNumbers,
+	const ImperatorWorld::World& sourceWorld) const
+{
+	// determine ownership by province development.
+	std::map<int, std::vector<std::shared_ptr<ImperatorWorld::Province>>> theClaims; // owner, offered province sources
+	std::map<int, int> theShares;														// owner, development
+	int winner = -1;
+	auto maxDev = -1;
+
+	for (auto ck2ProvinceID : impProvinceNumbers)
+	{
+		const auto& impProvince = sourceWorld.getProvinces().find(ck2ProvinceID);
+		if (impProvince == sourceWorld.getProvinces().end())
+		{
+			continue; // Broken mapping, or loaded a mod changing provinces without using it.
+		}
+		auto owner = impProvince->second->getOwner();
+		theClaims[owner].push_back(impProvince->second);
+		theShares[owner] = lround(impProvince->second->getBuildingsCount());
+	}
+	// Let's see who the lucky winner is.
+	for (const auto& share : theShares)
+	{
+		if (share.second > maxDev)
+		{
+			winner = share.first;
+			maxDev = share.second;
+		}
+	}
+	if (winner == -1)
+	{
+		return std::nullopt;
+	}
+
+	// Now that we have a winning owner, let's find its largest province to use as a source.
+	maxDev = -1; // We can have winning provinces with weight = 0;
+
+	std::pair<int, std::shared_ptr<ImperatorWorld::Province>> toReturn;
+	for (const auto& province : theClaims[winner])
+	{
+		const auto provinceWeight = province->getBuildingsCount();
```

Comment:
Buildings, rather than population?

Reply:
We should probably use both (will add to PR), or even more factors (maybe not in the initial release).